tests/DataSourceTests.js

```diff
@@ -15,61 +15,199 @@ https://github.com/GPII/kettle/LICENSE.txt
+        path: "/path",
+        method: "GET",
+        host: "thing.available:997", // appears to be a parsing fault in url module
+        family: 4,
+        headers: {
+            "x-custom-header": "x-custom-value",
+            "x-custom-header2": "x-custom-value2"
+        }
+    },
+    creatorArgs: {
+        port: 998,
+        family: 4
+    },
+    directOptions: {
+        headers: {
+            "x-custom-header2": "x-custom-value2"
+        },
+        port: 999
+    }
+}, {
+    gradeName: "kettle.tests.KETTLE34dataSource2",
+    expectedOptions: {
+        protocol: "http:",
+        port: 998,
+        path: "/device",
+        host: "localhost"
+    }
+}
+];
+
+kettle.tests.dataSource.resolutionTest = function (fixture) {
+    jqUnit.test("KETTLE-34 request option resolution test", function () {
+        var httpRequest = http.request,
+            capturedOptions;
+        http.request = function (requestOptions) {
+            capturedOptions = requestOptions;
+            return {
+                on: fluid.identity,
+                end: fluid.identity
+            };
+        };
+        try {
+            var dataSource = fluid.invokeGlobalFunction(fixture.gradeName, [fixture.creatorArgs]);
+            dataSource.get(null, fixture.directOptions);
+            jqUnit.assertLeftHand("Resolved expected requestOptions", fixture.expectedOptions, capturedOptions);
+        } finally {
+            http.request = httpRequest;
+        }
+    });
 };
 
+fluid.each(kettle.tests.dataSource.resolutionTests, function (fixture) {
+    kettle.tests.dataSource.resolutionTest(fixture);
+});
+
+
+
+
+fluid.defaults("kettle.tests.KETTLE38base", {
+    gradeNames: "kettle.dataSource.URL",
+    writable: true
+});
+
+fluid.defaults("kettle.tests.KETTLE38derived", {
+    gradeNames: "kettle.tests.KETTLE38base"
+});
+
+jqUnit.test("KETTLE-38 derived writable dataSource test", function () {
+    var dataSource = kettle.tests.KETTLE38derived();
+    jqUnit.assertValue("Resolved writable grade via base grade", dataSource.set);
+});
+
+
+kettle.tests.formencData = {
+    "text1": "text default",
+    "text2": "aωb"
+};
+
+fluid.defaults("kettle.tests.formencSource", {
+    gradeNames: "kettle.dataSource.file.moduleTerms",
+    path: "%kettle/tests/data/formenc.txt",
+    components: {
+        encoding: {
+            type: "kettle.dataSource.encoding.formenc"
+        }
+    }
+});
+
+jqUnit.asyncTest("Reading file in formenc encoding", function () {
+    var that = kettle.tests.formencSource();
+    jqUnit.expect(1);
+    that.get().then(function (result) {
+        jqUnit.assertDeepEq("Received decoded result", kettle.tests.formencData, result);
+        jqUnit.start();
+    }, function (err) {
+        jqUnit.fail("Got error " + JSON.stringify(err));
+    });
+});
+
+fluid.defaults("kettle.tests.formencSourceWrite", {
+    gradeNames: "kettle.dataSource.file.moduleTerms",
+    path: "%kettle/tests/data/writeable/formenc.txt",
```

Comment:
`DataSourceTests.js` fails unless the directory `kettle/tests/data/writeable` exists (at least on Linux):

```
jq: FAIL: Test name "Writing file in formenc encoding" - Message: Got error {"message":"ENOENT, open '/home/simon/www/fluid-project/kettle/tests/data/writeable/formenc.txt'"}
```

This directory does not exist in the project repository.